lib/http_aws_sigv4.c

```diff
@@ -120,214 +114,280 @@ CURLcode Curl_output_aws_sigv4(struct Curl_easy *data, bool proxy)
+    provider1_mid[len] = '\0';
+
+    if(tmp1) {
+      tmp0 = tmp1 + 1;
+      tmp1 = strchr(tmp0, ':');
+      len = tmp1 ? (size_t)(tmp1 - tmp0) : strlen(tmp0);
+      if(len < 1) {
+        infof(data, "region can't be empty\n");
+        ret = CURLE_BAD_FUNCTION_ARGUMENT;
+        goto fail;
+      }
+      region = Curl_memdup(tmp0, len + 1);
+      if(!region) {
+        goto fail;
+      }
+      region[len] = '\0';
+
+      if(tmp1) {
+        tmp0 = tmp1 + 1;
+        service = strdup(tmp0);
+        if(!service) {
+          goto fail;
+        }
+        if(strlen(service) < 1) {
+          infof(data, "service can't be empty\n");
+          ret = CURLE_BAD_FUNCTION_ARGUMENT;
+          goto fail;
+        }
+      }
     }
-    Curl_strntolower(low_provider, provider, provider_l);
-    Curl_strntolower(mid_provider, provider, provider_l);
-  }
-  else if(provider_l <= PROVIDER_MAX_L) {
-    Curl_strntolower(low_provider0, provider, provider_l);
-    Curl_strntolower(low_provider, provider, provider_l);
-    Curl_strntolower(mid_provider, provider, provider_l);
-    Curl_strntoupper(up_provider, provider, provider_l);
-    mid_provider[0] = Curl_raw_toupper(mid_provider[0]);
   }
   else {
-    infof(data, "v4 signature argument string too long\n");
-    return CURLE_BAD_FUNCTION_ARGUMENT;
+    provider1_low = Curl_memdup(provider0_low, len + 1);
+    provider1_mid = Curl_memdup(provider0_low, len + 1);
+    if(!provider1_low || !provider1_mid) {
+      goto fail;
+    }
+    provider1_mid[0] = Curl_raw_toupper(provider1_mid[0]);
+  }
+
+  if(!service) {
+    tmp0 = hostname;
+    tmp1 = strchr(tmp0, '.');
+    len = tmp1 - tmp0;
+    if(!tmp1 || len < 1) {
+      infof(data, "service missing in parameters or hostname\n");
+      ret = CURLE_URL_MALFORMAT;
+      goto fail;
+    }
+    service = Curl_memdup(tmp0, len + 1);
+    if(!service) {
+      goto fail;
+    }
+    service[len] = '\0';
+
+    if(!region) {
+      tmp0 = tmp1 + 1;
+      tmp1 = strchr(tmp0, '.');
+      len = tmp1 - tmp0;
+      if(!tmp1 || len < 1) {
+        infof(data, "region missing in parameters or hostname\n");
+        ret = CURLE_URL_MALFORMAT;
+        goto fail;
+      }
+      region = Curl_memdup(tmp0, len + 1);
+      if(!region) {
+        goto fail;
+      }
+      region[len] = '\0';
+    }
   }
 
 #ifdef DEBUGBUILD
   force_timestamp = getenv("CURL_FORCETIME");
   if(force_timestamp)
-    rawtime = 0;
+    clock = 0;
   else
+    time(&clock);
+#else
+  time(&clock);
 #endif
-    time(&rawtime);
-
-  ret = Curl_gmtime(rawtime, &info);
+  ret = Curl_gmtime(clock, &tm);
   if(ret != CURLE_OK) {
-    return ret;
+    goto fail;
   }
-
-  if(!strftime(date_iso, sizeof(date_iso), "%Y%m%dT%H%M%SZ", &info)) {
-    return CURLE_OUT_OF_MEMORY;
+  if(!strftime(timestamp, sizeof(timestamp), "%Y%m%dT%H%M%SZ", &tm)) {
+    goto fail;
   }
-
-  memcpy(date, date_iso, sizeof(date));
+  (void)memcpy(date, timestamp, sizeof(date));
```

Comment:
It seems I can't unresolved a conversation.
So I can understand why a project would adopt the "void cast unused function return" rule.
but curl didn't adopt that, and this line is (it seems) the only line in the whole libcurl source code that cast a function return.
So if you want to do that, maybe you should at first make a PR that change the coding style, and fix other usage in libcurl too ? maybe discuss this before to see if your change will be accepted.

Reply:
Where in the curl codestyle is it forbidden to format the code like this?

Reply:
it's not, but it's not ask either.
I'm just not sure it's a good idea to start using that style even though, it's used nowhere else in the code.
I mean doing that make sense only if you use that kind of style everywhere in the code. so you could use static analyser tools, but doing this on only a few functions, seems to me, to be just unnecessary noise.

Reply:
What do you think, @bagder?

Reply:
I count 219 instances of `memcpy()` in the libcurl code right now and not one of them are prefixed with `(void)`. I don't think it adds anything and consistency is good. I think you can remove it. The return code from memcpy doesn't help anyone.